docs/source/tutorial/convnet.rst

```diff
@@ -0,0 +1,428 @@
+
+ResNet152
+'''''''''
+
+How about ResNet? ResNet [He16]_ came in the next year's ILSVRC. It is way deeper
+model than VGG16, so the depth is up to 152 layers. This sounds super laboring
+to build, but it can be implemented in almost same manner as VGG16. In the
+other words, it's easy. One possible way to write ResNet-152 is:
+
+.. doctest::
+
+    class ResNet152(chainer.ChainList):
+
+        def __init__(self, n_blocks=[3, 8, 36, 3]):
+            super(ResNet152, self).__init__(
+                conv1=L.Convolution2D(
+                    None, 64, 7, 2, 3, initialW=w, nobias=True),
+                bn1=L.BatchNormalization(64),
+                res2=ResBlock(n_blocks[0], 64, 64, 256, 1),
+                res3=ResBlock(n_blocks[1], 256, 128, 512),
+                res4=ResBlock(n_blocks[2], 512, 256, 1024),
+                res5=ResBlock(n_blocks[3], 1024, 512, 2048))
+            self.train = True
+
+        def __call__(self, x):
+            h = self.bn1(self.conv1(x), test=not self.train)
+            h = F.max_pooling_2d(F.relu(h), 2, 2)
+            h = self.res2(h)
+            h = self.res3(h)
+            h = self.res4(h)
+            h = self.res5(h)
+            if self.train:
+                return h
+            return F.softmax(h)
+
+    class ResBlock(chainer.ChainList):
+
+        def __init__(self, n_layers, n_in, n_mid, n_out, stride=2):
+            w = chainer.initializers.HeNormal()
+            super(ResBlock, self).__init__()
+            self.add_link(BottleNeck(n_in, n_mid, n_out, stride, True))
+            for _ in range(n_layers - 1):
+                self.add_link(BottleNeck(n_out, n_mid, n_out))
+
+        def __call__(self, x, train):
+            for f in self.children():
+                x = f(x, train)
+            return x
+
+    class BottleNeck(chainer.Chain):
+
+        def __init__(self, n_in, n_mid, n_out, stride=1, proj=False):
+            w = chainer.initializers.HeNormal()
+            super(BottleNeck, self).__init__(
+                conv1x1a=L.Convolution2D(
+                    n_in, n_mid, 1, stride, 0, initialW=w, nobias=True),
+                conv3x3b=L.Convolution2D(
+                    n_mid, n_mid, 3, 1, 1, initialW=w, nobias=True),
+                conv1x1c=L.Convolution2D(
+                    n_mid, n_out, 1, 1, 0, initialW=w, nobias=True),
+                bn_a=L.BatchNormalization(n_mid),
+                bn_b=L.BatchNormalization(n_mid),
+                bn_c=L.BatchNormalization(n_out))
+            if proj:
+                self.add_link('conv1x1r', L.Convolution2D(
+                    n_in, n_out, 1, stride, 0, initialW=w, nobias=True))
+                self.add_link('bn_r', L.BatchNormalization(n_out))
+            self.proj = proj
+
+        def __call__(self, x, train):
+            h = F.relu(self.bn_a(self.conv1x1a(x), test=not train))
+            h = F.relu(self.bn_b(self.conv3x3b(x), test=not train))
+            h = self.bn_c(self.conv1x1c(x), test=not train)
```

Comment:
Is this code correct? It looks like you keep overwriting h because the value is not used in the following lines (it keeps reading from x only).